### PR TITLE
#3348 New Sensor Step Two

### LINKED ISSUE #3348: New Sensor Flow/Page
## Is your feature request related to a problem? Please describe.

The new sensor flow and page

## Describe the solution you'd like

As per the marvel prototype

## Implementation

Quite a big page with a lot of steps. Going to break it down into the following tasks:

- [ ] `Stepper`
- [ ] Make `IconButton` have a `Rectangle` variant?
- [ ] Make step1 component
- [ ] Make step2 component
- [ ] Make step3 component
- [ ] Logic to send bluetooth commands + save a new sensor

## Describe alternatives you've considered

N/A

## Additional context

Resources

- https://marvelapp.com/prototype/55aea8f/screen/75648024
image

- https://drive.google.com/file/d/1wLg12JAQyrgB6D3F_X0O9LBZwJs8PcJM/view?usp=sharing
image

- https://drive.google.com/file/d/15SsY4YCmhpSgiihU1Q7FpZMnqCUkZcIp/view?usp=sharing
image

- https://drive.google.com/file/d/15SsY4YCmhpSgiihU1Q7FpZMnqCUkZcIp/view?usp=sharing
image

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #3348 

## Change summary

- Adds step two!

## Testing

Internal??

### Related areas to think about

- Went a little against the interface design and added a 'back' button as well as the above/below around the temperature
- Really want the temps/durations to be editable..
- The 'always 2 dp' is because when you're holding down the +/- it gets really seizure-inducing
- Colors instead of just black?!.. heh
- Kind of think the 'paper' should be a consistent size througout all the tabs. Something to think about
- Navigation is a touch slow between tabs. Something to optimize when the last ones in.
image

### CODE REVIEW

File: src/reducers/NewSensorReducer.js [hunk 1]
```diff
@@ -2,6 +2,10 @@ import { NEW_SENSOR_ACTIONS, VACCINE_ACTIONS } from '../actions';
 
 const initialState = () => ({
   macAddress: '',
+  HOT_CUMULATIVE: { duration: 300, temperature: 8 },
+  COLD_CUMULATIVE: { duration: 300, temperature: 2 },
+  HOT_CONSECUTIVE: { duration: 300, temperature: 8 },
+  COLD_CONSECUTIVE: { duration: 300, temperature: 2 },
```
Comment (joshxg) [inline src/reducers/NewSensorReducer.js:8]: defaults for now? some level of customization of these in the future

File: src/reducers/VaccineReducer.js [hunk 2]
```diff
@@ -32,6 +32,7 @@ export const VaccineReducer = (state = initialState(), action) => {
     case VACCINE_ACTIONS.SCAN_STOP: {
       return {
         ...state,
+        scannedSensorAddresses: [],
```
Comment (joshxg) [inline src/reducers/VaccineReducer.js:35]: resetting this when stopping a scan also - had a bit of jank when navigating back to the tab

File: src/selectors/newSensor.js [hunk 3]
```diff
@@ -2,3 +2,55 @@ export const selectNewSensor = ({ newSensor }) => {
...
+  const hotConsecutiveConfigTempThreshold = selectHotConsecutiveTemperatureThreshold(state);
+
+  return {
+    COLD_CUMULATIVE: coldCumulativeConfigTempThreshold,
+    HOT_CUMULATIVE: hotCumulativeConfigTempThreshold,
+    COLD_CONSECUTIVE: coldConsecutiveConfigTempThreshold,
+
+    HOT_CONSECUTIVE: hotConsecutiveConfigTempThreshold,
+  };
+};
```
Comment (joshxg) [inline src/selectors/newSensor.js:56]: basically ensuring the sets of temperatures for each of the pairs: cumulative/consecutive are disjoint between hot/cold. Maybe a nicer way to do this but 🤷

File: src/widgets/StepperInputs/DurationEditor.js [hunk 4]
```diff
@@ -24,8 +24,8 @@ export const DurationEditor = ({
 }) => {
   const formatted = String(keepInRange(value, 1, 999));
 
-  const onIncrement = () => onChange(value + stepAmount);
-  const onDecrement = () => onChange(value - stepAmount);
+  const onIncrement = () => onChange(keepInRange(value + stepAmount, 1, 999));
```
Comment (joshxg) [inline src/widgets/StepperInputs/DurationEditor.js:27]: keeping the duration between 1,999.. not sure why

File: src/widgets/StepperInputs/TemperatureEditor.js [hunk 5]
```diff
@@ -22,13 +24,20 @@ export const TemperatureEditor = ({
...
+  const maximum = above ? aboveMaximum : threshold;
+
   const formatted = keepInRange(value, minimum, maximum).toFixed(2);
 
-  const onIncrement = () => onChange(value + stepAmount);
-  const onDecrement = () => onChange(value - stepAmount);
+  const onIncrement = () =>
+    onChange(keepInRange(temperature(value + stepAmount).temperature(), minimum, maximum));
+  const onDecrement = () =>
+    onChange(keepInRange(temperature(value - stepAmount).temperature(), minimum, maximum));
```
Comment (joshxg) [inline src/widgets/StepperInputs/TemperatureEditor.js:40]: continuation of the logic required from the selectors to ensure each configs temperatures don't cross over

File: src/pages/NewSensor/NewSensorStepTwo.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+import { PaperSection, FlexRow, PageButton, Spacer } from '../../widgets';
+
+import { SUSSOL_ORANGE, WHITE } from '../../globalStyles';
+import { vaccineStrings } from '../../localization';
+import { WizardActions } from '../../actions/WizardActions';
+import { goBack } from '../../navigation/actions';
+
+export const NewSensorStepTwoComponent = ({ nextTab, previousTab, exit }) => (
+  <TabContainer>
+    <PaperSection height={320} headerText={vaccineStrings.new_sensor_step_two_title}>
```
Comment (mark-prins) [inline src/pages/NewSensor/NewSensorStepTwo.js:17]: the fixed values for heights are necessary?

**Review by mark-prins — APPROVED** [review]
Looks good!
Nice to use too 🙂 
Strangely, my phone picks up 7 sensors, and this only finds 4 🤷‍♂️

File: src/pages/NewSensor/NewSensorPage.js [hunk 7]
```diff
@@ -4,17 +4,15 @@ import { View } from 'react-native';
...
-  <View style={{ backgroundColor: 'green', flex: 1, width: 500, height: 500 }} />
-);
 const TabThree = () => (
   <View style={{ backgroundColor: 'blue', flex: 1, width: 500, height: 500 }} />
 );
 
 const tabs = [
-  { component: NewSensorStepOne, name: 'prescription', title: '' },
-  { component: TabTwo, name: 'prescription', title: '' },
+  { component: NewSensorStepOne, name: 'newSensorStepThree', title: '' },
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/pages/NewSensor/NewSensorPage.js]: 9e033e1

File: src/pages/NewSensor/NewSensorStepTwo.js (hunk 6 above)
Comment (joshxg) [reply src/pages/NewSensor/NewSensorStepTwo.js:17]: If I say yes do I have to explain more..? Because I just know that no look good without 😅

Comment (mark-prins) [reply src/pages/NewSensor/NewSensorStepTwo.js:17]: ok, I'm convinced. that's a strong argument you've put forward there.

**Comment by joshxg** [comment]
I am a little concerned about not picking up the sensors. I'll look into it..

**Pull request merged**